Replace free() by munmap()

Xphalnos opened the pull request:
## Description
Changing `free` by `munmap` for a safest memory management.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [X] Refactoring (no functional change, code cleanup)
- [ ] Performance improvement

## Code Review Process
**IMPORTANT**: This project maintains an extremely high standard for code quality and correctness. By submitting this pull request, you acknowledge and agree to the following:

1. **You must be able to defend every single line of code you have added or modified**
2. **You must be prepared to explain the reasoning behind every design decision**
3. **You must be able to discuss and justify your approach to solving the problem**
4. **You must be ready to address all feedback and make requested changes**
5. **You must have thoroughly tested your changes and be confident in their correctness**

The code review process for this project is intentionally tedious and thorough. Do not submit a pull request unless you are prepared to defend your PR changes.

## Breaking Changes
If this change introduces any breaking changes, please describe them here.

## Checklist
- [X] My code follows the project's coding style guidelines
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have read the project's CONTRIBUTING guidelines
- [X] I understand and agree to the rigorous code review process described above

Xphalnos committed:
Replace free() by munmap()

Signed-off-by: Xphalnos <yiga.steam@gmail.com>

GloriousTacoo reviewed (changes requested):
This is very unsafe compared to free() because there is a lot that can go wrong and you've done zero error checking.

From munmap's manpage:
```
The munmap() function shall fail if:

       EINVAL The len argument is 0.

       The munmap() function may fail if:

       EINVAL The addr argument is not a multiple of the page size as
              returned by sysconf().

```
Add asserts to make this function impossible to fail. Then check the return value of munmap and if its still -1 add the assert `PVM_ASSERT_MSG(false, "Failed to free arena memory");`

GloriousTacoo commented:
You did the second half of my request but not the first:

> Add asserts to make this function impossible to fail.

I will work on this PR over the weekend.

GloriousTacoo commented:
By the way I'll use this to show what I mean:
```cpp
#src/jit/decoder/arm32.cpp

void arm32_add_instruction(arm32_decoder_t* decoder, const char* name, const char* bitstring,
 arm32_handler_fn handler)                                                                            {
     PVM_ASSERT(nullptr != decoder);
     PVM_ASSERT(nullptr != decoder->allocator.data);
     PVM_ASSERT(nullptr != name);
     PVM_ASSERT(nullptr != bitstring);
     PVM_ASSERT(decoder->instruction_count < decoder->instruction_capacity);
     ...
```
All those asserts at the beginning guarantees that the function will not fail when it is called. I will do the same for `munmap`.

GloriousTacoo committed:
host: add assertions

Munmap can fail if the length argument is 0, and the address being freed
is not a multiple of the host's page size.

Signed-off-by: Ronald Caesar <github43132@proton.me>
